Add timings to vnode_get/put stats

russelldb opened the pull request:
Add per partition stats for get/put. Test as with the other stat PRs. Do some gets / puts against a node, attach to it and run:

```
[{Stat, folsom_metrics:get_metric_value(Stat)} || 
 Stat <- folsom_metrics:get_metrics(), 
             is_tuple(Stat),
             element(1, Stat) == riak_kv,
             element(2, Stat) == vnode_gets].  
```

Or

```
[{Stat, folsom_metrics:get_metric_value(Stat)} || 
 Stat <- folsom_metrics:get_metrics(), 
             is_tuple(Stat),
             element(1, Stat) == riak_kv,
             element(2, Stat) == vnode_puts].  
```

russelldb committed:
Add timings to vnode_get/put stats

Add per partition stats for get/put.

russelldb committed:
Use riak core stat query for riak_kv_stat get_stats

Move backwards compatiblity code to own module for easier
deprecation and removal.

russelldb committed:
Add new read repair dimensioned stat aggregations to stat blob

russelldb committed:
Add pipe stats to THE BLOB

coderoshi commented:
+1

beerriot commented on src/riak_kv_vnode.erl:
```diff
@@ -280,9 +280,11 @@ handle_command(?KV_PUT_REQ{bkey=BKey,
                            options=Options},
                Sender, State=#state{idx=Idx}) ->
+    StartTS = os:timestamp(),
     riak_kv_mapred_cache:eject(BKey),
```
A note to PR #433: merging with this commit may create a conflict that tries to resurrect the call to `riak_kv_mapred_cache`.

russelldb replied:
Noted, thank you.

beerriot commented:
It's quite interesting that per-vnode stats are not expunged when the vnode shuts down. I think that's a good thing, because you won't "just miss" a vnode's final stats after it hands off. And, given riak's on-demand vnode startup, "zero" spirals for non-running vnodes are technically correct.

It might be a good idea to head off some possible worry and questions by explaining this in our user docs, though: why some nodes show stats for all vnodes, others only some; why it's okay that many show zero forever; etc.

russelldb commented:
Good point. I don't believe we have _any_ docs about stats. The release notes is one place that this can be mentioned. I'll make sure it is (once this PR is merged.)

That aside, is this a +1?

beerriot commented:
Sounds great, +1.

russelldb committed:
Use binary bucket names, rather than converting to atoms

Technically there is no bound to the number of buckets one can
create, and no limit the length of a bucket name. The erlang atom
table is a finite resource and atom sapce is not reclaimed so
creating unbounded atoms of any length is a bad idea.

russelldb committed:
Address review comments

Re-name and better comment confusing `recursive_join` fun.
Remove redundant call to `lists:flatten` as once is enough.
Correct markdown style escapes to edoc style.
